cdr/resolver.go

```diff
@@ -0,0 +1,247 @@
+// the same (as far as I can tell the Google DNS implementation doesn't share
+// cache state between the distributed nodes so this should be safe).
+//
+// Since DNS isn't a super secure protocol and Google has recently introduced
+// a public HTTPS API for their DNS resolver so instead we use that.
+//
+// API reference:
+//   https://developers.google.com/speed/public-dns/docs/dns-over-https#api_specification
+
+var apiURI = "https://dns.google.com/resolve"
+
+func parseAnswer(as []core.GPDNSAnswer) ([]*dns.CAA, error) {
+	rrs := []*dns.CAA{}
+	// only bother parsing out CAA records
+	for _, a := range as {
+		if a.Type != dns.TypeCAA {
+			continue
+		}
+		rr, err := dns.NewRR(fmt.Sprintf("%s %d IN CAA %s", a.Name, a.TTL, a.Data))
+		if err != nil {
+			return nil, err
+		}
+		if caaRR, ok := rr.(*dns.CAA); ok {
+			rrs = append(rrs, caaRR)
+		}
+	}
+	return rrs, nil
+}
+
+func createClient(proxy string) (*http.Client, string, error) {
+	u, err := url.Parse(proxy)
+	if err != nil {
+		return nil, "", err
+	}
+	transport := &http.Transport{
+		Proxy:               http.ProxyURL(u),
+		TLSHandshakeTimeout: 10 * time.Second, // Same as http.DefaultTransport, doesn't override context
+	}
+	return &http.Client{
+		Transport: transport,
+	}, u.Host, nil
+}
+
+// CAADistributedResolver holds state needed to talk to GPDNS
+type CAADistributedResolver struct {
+	URI         string
+	Clients     map[string]*http.Client
+	stats       metrics.Scope
+	maxFailures int
+	timeout     time.Duration
+	logger      blog.Logger
+}
+
+// New returns an initialized CAADistributedResolver which requires a M of N
+// quorum to succeed where N = len(proxies) and M = N - maxFailures
+func New(scope metrics.Scope, timeout time.Duration, maxFailures int, proxies []string, logger blog.Logger) (*CAADistributedResolver, error) {
+	cdr := &CAADistributedResolver{
+		Clients:     make(map[string]*http.Client, len(proxies)),
+		URI:         apiURI,
+		stats:       scope,
+		maxFailures: maxFailures,
+		timeout:     timeout,
+		logger:      logger,
+	}
+	for _, p := range proxies {
+		c, h, err := createClient(p)
+		if err != nil {
+			return nil, err
+		}
+		cdr.Clients[h] = c
+	}
+	return cdr, nil
+}
+
+// queryCAA sends the query request to the GPD API. If the return code is
+// dns.RcodeSuccess the 'Answer' section is parsed for CAA records, otherwise
+// a error is returned. Unlike bdns.DNSResolver.LookupCAA it will not repeat
+// failed queries if the context has not expired as we expect to be running
+// multiple queries in parallel and only need a M of N quorum (we also expect
+// GPD to have quite good availability)
+func (cdr *CAADistributedResolver) queryCAA(ctx context.Context, req *http.Request, ic *http.Client) ([]*dns.CAA, error) {
+	apiResp, err := ctxhttp.Do(ctx, ic, req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		_ = apiResp.Body.Close()
+	}()
+	body, err := ioutil.ReadAll(apiResp.Body)
```

Comment:
I trust Google would never have a serious error in their implementation like this, but just for safety's sake, I'd feel better if we avoided `ReadAll` for uncontrolled external loads like this. In theory, this could hang while we download some enormous file.